M1 iOS simulator app fails to build with `ld: framework not found Pods_Runner` `Undefined symbols for architecture x86_64`

<!-- Thank you for using Flutter!

     If you are looking for support, please check out our documentation
     or consider asking a question on Stack Overflow:
      * https://flutter.dev/
      * https://api.flutter.dev/
      * https://stackoverflow.com/questions/tagged/flutter?sort=frequent

     If you have found a bug or if our documentation doesn't have an answer
     to what you're looking for, then fill out the template below. Please read
     our guide to filing a bug first: https://flutter.dev/docs/resources/bug-reports
-->

## Steps to Reproduce

<!-- Please include full steps to reproduce so that we can reproduce the problem. -->

Upgrade existing projects on M1 Macs, then try to build the iOS target, flutter will throw errors with generatedplugincontraints.h module not found error. 
initially this was with cloud_fire_store, then I removed firebase from the app, and it is throwing the same error with device info.
So for whatever packages, if it is listed as the first package in the generatedpluginconstraint.h file, I will get this module not found issue.

I even tried deleting the whole iOS folder, and regenerate it with `flutter create .`, same issue.

I tried to build the same project on an intel based app with Flutter 2.8.0, works perfectly fine.

I tried to build the macOS target on M1 Macs with Flutter 2.8.0, works fine too.

I then rolled back to version 2.5.3 on M1 Macs, build the iOS target for same project, works perfectly fine as well.

So at this point I think it maybe a Flutter bug itself, whenever I have a native side plugin in my project it will not compile the iOS target on Apple Silicon Macs.

**Expected results:** <!-- what did you want to see? -->

**Actual results:** <!-- what did you see? -->

<details>
<summary>Code sample</summary>

<!--
      Please create a minimal reproducible sample that shows the problem
      and attach it below between the lines with the backticks.

      To create it you can use `flutter create bug` command and update the `main.dart` file.

      Alternatively, you can use https://dartpad.dev/
      which is capable of creating and running small Flutter apps.

      Without this we will unlikely be able to progress on the issue, and because of that
      we regretfully will have to close it.
-->

```dart
```

</details>

<details>
  <summary>Logs</summary>

<!--
      Run your application with `flutter run --verbose` and attach all the
      log output below between the lines with the backticks. If there is an
      exception, please see if the error message includes enough information
      to explain how to solve the issue.
-->

```
```

<!--
     Run `flutter analyze` and attach any output of that command below.
     If there are any analysis errors, try resolving them before filing this issue.
-->

```
Doctor summary (to see all details, run flutter doctor -v):
[✓] Flutter (Channel stable, 2.8.0, on macOS 12.0.1 21A559 darwin-arm, locale en-AU)
[✓] Android toolchain - develop for Android devices (Android SDK version 31.0.0)
[✓] Xcode - develop for iOS and macOS (Xcode 13.1)
[✓] Chrome - develop for the web
[✓] Android Studio (version 2020.3)
[✓] IntelliJ IDEA Ultimate Edition (version 2021.2.3)
[✓] IntelliJ IDEA Ultimate Edition (version 2021.2.3)
[✓] VS Code (version 1.63.0)
[✓] Connected device (3 available)

• No issues found!
```

<!-- Finally, paste the output of running `flutter doctor -v` here. -->


```
Running "flutter pub get" in upstreet_mobile_app...
Launching lib/main.dart on iPhone 13 in debug mode...
Xcode build done.                                           19.1s
Failed to build iOS app
Error output from Xcode build:
↳
    ** BUILD FAILED **
Xcode's output:
↳
    ld: warning: ignoring file /Users/hongfeiyang/upstreet/upstreet_mobile_app/build/ios/Debug-iphonesimulator/AppAuth/AppAuth.framework/AppAuth, building for iOS Simulator-x86_64 but attempting to link with file built for iOS Simulator-arm64
    ld: warning: ignoring file /Users/hongfeiyang/upstreet/upstreet_mobile_app/build/ios/Debug-iphonesimulator/GTMAppAuth/GTMAppAuth.framework/GTMAppAuth, building for iOS Simulator-x86_64 but attempting to link with file built for iOS Simulator-arm64
    ld: warning: ignoring file /Users/hongfeiyang/upstreet/upstreet_mobile_app/build/ios/Debug-iphonesimulator/GTMSessionFetcher/GTMSessionFetcher.framework/GTMSessionFetcher, building for iOS Simulator-x86_64 but attempting to link with file built for iOS Simulator-arm64
    Undefined symbols for architecture x86_64:
      "_OBJC_CLASS_$_GTMOAuth2KeychainCompatibility", referenced from:
          objc-class-ref in GoogleSignIn(GIDAuthStateMigration_bdf842e92d17b12b5f020e0e721bd017.o)
      "_OBJC_CLASS_$_OIDURLQueryComponent", referenced from:
          objc-class-ref in GoogleSignIn(GIDSignIn_f05654be6d1b2af59ab6df426a0c4eb9.o)
      "_OBJC_CLASS_$_OIDServiceConfiguration", referenced from:
          objc-class-ref in GoogleSignIn(GIDSignIn_f05654be6d1b2af59ab6df426a0c4eb9.o)
      "_OIDResponseTypeCode", referenced from:
          -[GIDSignIn authenticateInteractivelyWithOptions:] in GoogleSignIn(GIDSignIn_f05654be6d1b2af59ab6df426a0c4eb9.o)
      "_OBJC_CLASS_$_OIDAuthorizationService", referenced from:
          objc-class-ref in GoogleSignIn(GIDAuthentication_7b83cc27af8929ec7e4c9760d73a0f31.o)
          objc-class-ref in GoogleSignIn(GIDSignIn_f05654be6d1b2af59ab6df426a0c4eb9.o)
      "_OBJC_CLASS_$_GTMSessionFetcher", referenced from:
          objc-class-ref in GoogleSignIn(GIDSignIn_f05654be6d1b2af59ab6df426a0c4eb9.o)
      "_OBJC_CLASS_$_OIDAuthorizationRequest", referenced from:
          objc-class-ref in GoogleSignIn(GIDSignIn_f05654be6d1b2af59ab6df426a0c4eb9.o)
      "_OBJC_CLASS_$_GTMAppAuthFetcherAuthorization", referenced from:
          _OBJC_CLASS_$_GTMAppAuthFetcherAuthorizationWithEMMSupport in GoogleSignIn(GIDAuthentication_7b83cc27af8929ec7e4c9760d73a0f31.o)
          objc-class-ref in GoogleSignIn(GIDAuthentication_7b83cc27af8929ec7e4c9760d73a0f31.o)
          objc-class-ref in GoogleSignIn(GIDSignIn_f05654be6d1b2af59ab6df426a0c4eb9.o)
          objc-class-ref in GoogleSignIn(GIDAuthStateMigration_bdf842e92d17b12b5f020e0e721bd017.o)
         (maybe you meant: _OBJC_CLASS_$_GTMAppAuthFetcherAuthorizationWithEMMSupport, _OBJC_CLASS_$_GTMAppAuthFetcherAuthorizationEMMChainedDelegate )
      "_OIDOAuthTokenErrorDomain", referenced from:
          ___46-[GIDAuthentication refreshTokensWithHandler:]_block_invoke in GoogleSignIn(GIDAuthentication_7b83cc27af8929ec7e4c9760d73a0f31.o)
      "_OBJC_CLASS_$_OIDIDToken", referenced from:
          objc-class-ref in GoogleSignIn(GIDGoogleUser_2bdd4a676c033bc6ca21970e75d9159c.o)
          objc-class-ref in GoogleSignIn(GIDAuthentication_7b83cc27af8929ec7e4c9760d73a0f31.o)
          objc-class-ref in GoogleSignIn(GIDSignIn_f05654be6d1b2af59ab6df426a0c4eb9.o)
      "_OBJC_METACLASS_$_GTMAppAuthFetcherAuthorization", referenced from:
          _OBJC_METACLASS_$_GTMAppAuthFetcherAuthorizationWithEMMSupport in GoogleSignIn(GIDAuthentication_7b83cc27af8929ec7e4c9760d73a0f31.o)
         (maybe you meant: _OBJC_METACLASS_$_GTMAppAuthFetcherAuthorizationWithEMMSupport, _OBJC_METACLASS_$_GTMAppAuthFetcherAuthorizationEMMChainedDelegate )
      "_OIDOAuthErrorResponseErrorKey", referenced from:
          +[GIDAuthentication handleTokenFetchEMMError:completion:] in GoogleSignIn(GIDAuthentication_7b83cc27af8929ec7e4c9760d73a0f31.o)
      "_OBJC_CLASS_$_GTMKeychain", referenced from:
          objc-class-ref in GoogleSignIn(GIDAuthStateMigration_bdf842e92d17b12b5f020e0e721bd017.o)
      "_OBJC_CLASS_$_OIDAuthState", referenced from:
          objc-class-ref in GoogleSignIn(GIDAuthentication_7b83cc27af8929ec7e4c9760d73a0f31.o)
          objc-class-ref in GoogleSignIn(GIDSignIn_f05654be6d1b2af59ab6df426a0c4eb9.o)
    ld: symbol(s) not found for architecture x86_64
    clang: error: linker command failed with exit code 1 (use -v to see invocation)
    note: Using new build system
    note: Planning
    note: Build preparation complete
    note: Building targets in parallel
    /Users/hongfeiyang/upstreet/upstreet_mobile_app/ios/Pods/Pods.xcodeproj: warning: The iOS Simulator deployment target 'IPHONEOS_DEPLOYMENT_TARGET' is set to 8.0, but the range of supported deployment target versions is 9.0 to 15.0.99. (in target 'AppAuth' from project 'Pods')
    /Users/hongfeiyang/upstreet/upstreet_mobile_app/ios/Pods/Pods.xcodeproj: warning: The iOS Simulator deployment target 'IPHONEOS_DEPLOYMENT_TARGET' is set to 8.0, but the range of supported deployment target versions is 9.0 to 15.0.99. (in target 'GoogleSignIn' from project 'Pods')
Could not build the application for the simulator.
Error launching application on iPhone 13.
Exited (sigterm)

```

</details>


<!--
      Consider also attaching screenshots and/or videos to better
      illustrate the issue.

      You can upload them directly on GitHub.
      Beware that video file size is limited to 10MB.
-->

Hi @masewo, I seem to be getting a different error than you are. 
Are the logs you shared from before or after you tried https://github.com/flutter/flutter/issues/93908#issuecomment-978112668?

<details>
<summary>logs</summary>

```console
Undefined symbols for architecture x86_64:
                 "_OBJC_CLASS_$_GTMOAuth2KeychainCompatibility", referenced from:
                     objc-class-ref in GoogleSignIn(GIDAuthStateMigration_bdf842e92d17b12b5f020e0e721bd017.o)
                 "_OBJC_CLASS_$_OIDURLQueryComponent", referenced from:
                     objc-class-ref in GoogleSignIn(GIDSignIn_f05654be6d1b2af59ab6df426a0c4eb9.o)
                 "_OBJC_CLASS_$_OIDServiceConfiguration", referenced from:
                     objc-class-ref in GoogleSignIn(GIDSignIn_f05654be6d1b2af59ab6df426a0c4eb9.o)
                 "_OIDResponseTypeCode", referenced from:
                     -[GIDSignIn authenticateInteractivelyWithOptions:] in GoogleSignIn(GIDSignIn_f05654be6d1b2af59ab6df426a0c4eb9.o)
                 "_OBJC_CLASS_$_OIDAuthorizationService", referenced from:
                     objc-class-ref in GoogleSignIn(GIDAuthentication_7b83cc27af8929ec7e4c9760d73a0f31.o)
                     objc-class-ref in GoogleSignIn(GIDSignIn_f05654be6d1b2af59ab6df426a0c4eb9.o)
                 "_OBJC_CLASS_$_GTMSessionFetcher", referenced from:
                     objc-class-ref in GoogleSignIn(GIDSignIn_f05654be6d1b2af59ab6df426a0c4eb9.o)
                 "_OBJC_CLASS_$_OIDAuthorizationRequest", referenced from:
                     objc-class-ref in GoogleSignIn(GIDSignIn_f05654be6d1b2af59ab6df426a0c4eb9.o)
                 "_OBJC_CLASS_$_GTMAppAuthFetcherAuthorization", referenced from:
                     _OBJC_CLASS_$_GTMAppAuthFetcherAuthorizationWithEMMSupport in GoogleSignIn(GIDAuthentication_7b83cc27af8929ec7e4c9760d73a0f31.o)
                     objc-class-ref in GoogleSignIn(GIDAuthentication_7b83cc27af8929ec7e4c9760d73a0f31.o)
                     objc-class-ref in GoogleSignIn(GIDSignIn_f05654be6d1b2af59ab6df426a0c4eb9.o)
                     objc-class-ref in GoogleSignIn(GIDAuthStateMigration_bdf842e92d17b12b5f020e0e721bd017.o)
                    (maybe you meant: _OBJC_CLASS_$_GTMAppAuthFetcherAuthorizationEMMChainedDelegate, _OBJC_CLASS_$_GTMAppAuthFetcherAuthorizationWithEMMSupport )
                 "_OIDOAuthTokenErrorDomain", referenced from:
                     ___46-[GIDAuthentication refreshTokensWithHandler:]_block_invoke in GoogleSignIn(GIDAuthentication_7b83cc27af8929ec7e4c9760d73a0f31.o)
                 "_OBJC_CLASS_$_OIDIDToken", referenced from:
                     objc-class-ref in GoogleSignIn(GIDGoogleUser_2bdd4a676c033bc6ca21970e75d9159c.o)
                     objc-class-ref in GoogleSignIn(GIDAuthentication_7b83cc27af8929ec7e4c9760d73a0f31.o)
                     objc-class-ref in GoogleSignIn(GIDSignIn_f05654be6d1b2af59ab6df426a0c4eb9.o)
                 "_OBJC_METACLASS_$_GTMAppAuthFetcherAuthorization", referenced from:
                     _OBJC_METACLASS_$_GTMAppAuthFetcherAuthorizationWithEMMSupport in GoogleSignIn(GIDAuthentication_7b83cc27af8929ec7e4c9760d73a0f31.o)
                    (maybe you meant: _OBJC_METACLASS_$_GTMAppAuthFetcherAuthorizationEMMChainedDelegate, _OBJC_METACLASS_$_GTMAppAuthFetcherAuthorizationWithEMMSupport )
                 "_OIDOAuthErrorResponseErrorKey", referenced from:
                     +[GIDAuthentication handleTokenFetchEMMError:completion:] in GoogleSignIn(GIDAuthentication_7b83cc27af8929ec7e4c9760d73a0f31.o)
                 "_OBJC_CLASS_$_GTMKeychain", referenced from:
                     objc-class-ref in GoogleSignIn(GIDAuthStateMigration_bdf842e92d17b12b5f020e0e721bd017.o)
                 "_OBJC_CLASS_$_OIDAuthState", referenced from:
                     objc-class-ref in GoogleSignIn(GIDAuthentication_7b83cc27af8929ec7e4c9760d73a0f31.o)
                     objc-class-ref in GoogleSignIn(GIDSignIn_f05654be6d1b2af59ab6df426a0c4eb9.o)
               ld: symbol(s) not found for architecture x86_64
```
</details>

--- a/packages/flutter_tools/bin/podhelper.rb
+++ b/packages/flutter_tools/bin/podhelper.rb
@@ -52,6 +52,11 @@ def flutter_additional_ios_build_settings(target)
   release_framework_dir = File.expand_path(File.join(artifacts_dir, 'ios-release', 'Flutter.xcframework'), __FILE__)
 
   target.build_configurations.each do |build_configuration|
+    # Build both x86_64 and arm64 simulator archs for all dependencies. If a single plugin does not support arm64 simulators,
+    # the app and all frameworks will fall back to x86_64. Unfortunately that case is not detectable in this script.
+    # Therefore all pods must have a x86_64 slice available, or linking a x86_64 app will fail.
+    build_configuration.build_settings['ONLY_ACTIVE_ARCH'] = 'NO' if build_configuration.type == :debug
+
     # Profile can't be derived from the CocoaPods build configuration. Use release framework (for linking only).
     configuration_engine_dir = build_configuration.type == :debug ? debug_framework_dir : release_framework_dir
     Dir.new(configuration_engine_dir).each_child do |xcframework_file|

--- a/packages/flutter_tools/bin/podhelper.rb
+++ b/packages/flutter_tools/bin/podhelper.rb
@@ -33,9 +33,6 @@ end
 def flutter_additional_ios_build_settings(target)
   return unless target.platform_name == :ios
 
-  # Return if it's not a Flutter plugin (transitive dependency).
-  return unless target.dependencies.any? { |dependency| dependency.name == 'Flutter' }
-
   # [target.deployment_target] is a [String] formatted as "8.0".
   inherit_deployment_target = target.deployment_target[/\d+/].to_i < 9
 
@@ -56,6 +53,9 @@ def flutter_additional_ios_build_settings(target)
     # the app and all frameworks will fall back to x86_64. Unfortunately that case is not detectable in this script.
     # Therefore all pods must have a x86_64 slice available, or linking a x86_64 app will fail.
     build_configuration.build_settings['ONLY_ACTIVE_ARCH'] = 'NO' if build_configuration.type == :debug
+
+    # Skip other updates if it's not a Flutter plugin (transitive dependency).
+    next unless target.dependencies.any? { |dependency| dependency.name == 'Flutter' }
 
     # Profile can't be derived from the CocoaPods build configuration. Use release framework (for linking only).
     configuration_engine_dir = build_configuration.type == :debug ? debug_framework_dir : release_framework_dir

--- a/packages/flutter_tools/test/integration.shard/ios_content_validation_test.dart
+++ b/packages/flutter_tools/test/integration.shard/ios_content_validation_test.dart
@@ -16,6 +16,7 @@ import 'test_utils.dart';
 void main() {
   group('iOS app validation', () {
     String flutterRoot;
+    Directory pluginRoot;
     String projectRoot;
     String flutterBin;
     Directory tempDir;
@@ -29,18 +30,20 @@ void main() {
         'flutter',
       );
 
+      // Test a plugin example app to allow plugins validation.
       processManager.runSync(<String>[
         flutterBin,
         ...getLocalEngineArguments(),
         'create',
         '--verbose',
         '--platforms=ios',
-        '-i',
-        'objc',
+        '-t',
+        'plugin',
         'hello',
       ], workingDirectory: tempDir.path);
 
-      projectRoot = tempDir.childDirectory('hello').path;
+      pluginRoot = tempDir.childDirectory('hello');
+      projectRoot = pluginRoot.childDirectory('example').path;
     });
 
     tearDownAll(() {
@@ -56,6 +59,7 @@ void main() {
         File outputFlutterFrameworkBinary;
         Directory outputAppFramework;
         File outputAppFrameworkBinary;
+        File outputPluginFrameworkBinary;
 
         setUpAll(() {
           processManager.runSync(<String>[
@@ -85,11 +89,13 @@ void main() {
 
           outputAppFramework = frameworkDirectory.childDirectory('App.framework');
           outputAppFrameworkBinary = outputAppFramework.childFile('App');
+
+          outputPluginFrameworkBinary = frameworkDirectory.childDirectory('hello.framework').childFile('hello');
         });
 
         testWithoutContext('flutter build ios builds a valid app', () {
-          // Should only contain Flutter.framework and App.framework.
-          expect(frameworkDirectory.listSync().length, 2);
+          expect(outputPluginFrameworkBinary, exists);
+
           expect(outputAppFrameworkBinary, exists);
           expect(outputAppFramework.childFile('Info.plist'), exists);
 
@@ -202,16 +208,60 @@ void main() {
         }, skip: !platform.isMacOS || buildMode != BuildMode.release); // [intended] only makes sense on macos.
 
         testWithoutContext('validate obfuscation', () {
-          final ProcessResult grepResult = processManager.runSync(<String>[
+          // HelloPlugin class is present in project.
+          ProcessResult grepResult = processManager.runSync(<String>[
             'grep',
-            '-i',
-            'hello',
+            '-r',
+            'HelloPlugin',
+            pluginRoot.path,
+          ]);
+          // Matches exits 0.
+          expect(grepResult.exitCode, 0);
+
+          // Not present in binary.
+          grepResult = processManager.runSync(<String>[
+            'grep',
+            'HelloPlugin',
             outputAppFrameworkBinary.path,
           ]);
-          expect(grepResult.stdout, isNot(contains('matches')));
+          // Does not match exits 1.
+          expect(grepResult.exitCode, 1);
         });
       });
     }
+
+    testWithoutContext('builds all plugin architectures for simulator', () {
+      final ProcessResult buildSimulator = processManager.runSync(
+        <String>[
+          flutterBin,
+          ...getLocalEngineArguments(),
+          'build',
+          'ios',
+          '--simulator',
+          '--verbose',
+          '--no-codesign',
+        ],
+        workingDirectory: projectRoot,
+      );
+      expect(buildSimulator.exitCode, 0);
+
+      final File pluginFrameworkBinary = fileSystem.file(fileSystem.path.join(
+        projectRoot,
+        'build',
+        'ios',
+        'iphonesimulator',
+        'Runner.app',
+        'Frameworks',
+        'hello.framework',
+        'hello',
+      ));
+      expect(pluginFrameworkBinary, exists);
+      final ProcessResult archs = processManager.runSync(
+        <String>['file', pluginFrameworkBinary.path],
+      );
+      expect(archs.stdout, contains('Mach-O 64-bit dynamically linked shared library x86_64'));
+      expect(archs.stdout, contains('Mach-O 64-bit dynamically linked shared library arm64'));
+    });
 
     testWithoutContext('build for simulator with all available architectures', () {
       final ProcessResult buildSimulator = processManager.runSync(
@@ -250,6 +300,6 @@ void main() {
       expect(archs.stdout, contains('Mach-O 64-bit dynamically linked shared library arm64'));
     });
   }, skip: !platform.isMacOS, // [intended] only makes sense for macos platform.
-     timeout: const Timeout(Duration(minutes: 5))
+     timeout: const Timeout(Duration(minutes: 7))
   );
 }

